# Challenge 01: From Spreadsheet Chaos to Structured Database

# STAGE 1 — Problem Overview
Retail companies often store operational data across multiple spreadsheets created by different departments. These spreadsheets are inconsistent in format, contain missing values, duplicate entries, and lack relationships between datasets.

This spreadsheet chaos makes it difficult to analyze sales performance, understand customer behavior, track store efficiency, and make reliable business decisions.

### Goal
The goal of this project is to transform messy and inconsistent spreadsheet data into a structured relational database. This will improve data quality, enable efficient querying, and support business intelligence, analytics, and forecasting.

### Key Challenges
• Inconsistent column names across files  
• Missing values and incomplete records  
• Duplicate entries  
• Lack of relationships between tables  
• Mixed data types and formatting issues  
• Data scattered across multiple files

### Proposed Approach
To solve these challenges, the project will follow a structured approach:

1. Understand and explore raw datasets
2. Clean and standardize data formats
3. Create a comprehensive data dictionary
4. Design an Entity-Relationship Diagram (ERD)
5. Build a normalized relational database schema
6. Develop ETL pipelines to clean and load data
7. Perform exploratory data analysis (EDA)
8. Build a baseline predictive model

### Expected Outcome
By the end of this project, we will develop a clean, normalized, and well-documented relational database that enables reliable analytics, reporting, and machine learning workflows. This structured foundation will support better business insights and decision-making.

### Dataset Overview
The dataset contains 20 files including 7 CSV files that store the primary retail data. Across these files, there are 158 columns representing various business attributes such as products, stores, transactions, and customer details.

The data is fragmented and spread across multiple sources, reflecting real-world spreadsheet chaos that requires systematic structuring and normalization.

In [1]:
import os

In [2]:
%pwd

'c:\\Users\\HP\\Desktop\\Python\\Vs_Python\\Data analysis\\Kaggle_Competition\\Notebook'

In [3]:
os.chdir("../")

In [4]:
%pwd

'c:\\Users\\HP\\Desktop\\Python\\Vs_Python\\Data analysis\\Kaggle_Competition'

In [5]:
import pandas as pd
pd.set_option('display.max_columns', None)

# STAGE 2 - Data Loading..

In [39]:
inventory = pd.read_csv(r'data\raw\inventory_snapshots_2024.csv')
product = pd.read_csv(r'data\raw\products_2024.csv')
promotion = pd.read_csv(r'data\raw\promotions_2024.csv')
sale_q1 = pd.read_csv(r'data\raw\sales_2024_q1.csv')
sale_q2 = pd.read_csv(r'data\raw\sales_2024_q2.csv')
sale_q3 = pd.read_csv(r'data\raw\sales_2024_q3.csv')
sale_q4 = pd.read_csv(r'data\raw\sales_2024_q4.csv')

# SATGE 3 - Data Understanding

### Inventory

In [40]:
inventory.head()

,snapshot_date,store_id,store_name,store_address,store_city,store_state,store_zip,product_upc,product_name,department_name,on_hand_qty,unit_cost,inventory_cost_value
0,2024-01-01,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,433218196001,Rainshadow Produce Apples,Produce,35,2.31,80.85
1,2024-01-01,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,386379402654,Evergreen Farms Bananas,Produce,43,3.68,158.24
2,2024-01-01,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,161559407816,Evergreen Farms Kale,Produce,42,1.10,46.20
3,2024-01-01,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,931034131647,Rainshadow Produce Spinach,Produce,20,0.84,16.80
4,2024-01-01,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,534192832764,North Fork Organics Carrots,Produce,55,3.19,175.45


In [41]:
inventory.shape

(2160, 13)

In [42]:
inventory.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   snapshot_date         2160 non-null   object 
 1   store_id              2160 non-null   int64  
 2   store_name            2160 non-null   object 
 3   store_address         2160 non-null   object 
 4   store_city            2160 non-null   object 
 5   store_state           2160 non-null   object 
 6   store_zip             2160 non-null   int64  
 7   product_upc           2160 non-null   int64  
 8   product_name          2160 non-null   object 
 9   department_name       2160 non-null   object 
 10  on_hand_qty           2160 non-null   int64  
 11  unit_cost             2160 non-null   float64
 12  inventory_cost_value  2160 non-null   float64
dtypes: float64(2), int64(4), object(7)
memory usage: 219.5+ KB


In [43]:
inventory.isnull().sum()

snapshot_date           0
store_id                0
store_name              0
store_address           0
store_city              0
store_state             0
store_zip               0
product_upc             0
product_name            0
department_name         0
on_hand_qty             0
unit_cost               0
inventory_cost_value    0
dtype: int64

In [44]:
inventory.duplicated().sum()

np.int64(0)

In [45]:
inventory.describe()

,store_id,store_zip,product_upc,on_hand_qty,unit_cost,inventory_cost_value
count,2160.000000,2160.000000,2.160000e+03,2160.000000,2160.000000,2160.000000
mean,102.000000,98112.000000,5.219156e+11,50.350000,6.306500,322.039241
std,0.816686,0.816686,2.887385e+11,19.694722,5.559481,330.897856
min,101.000000,98111.000000,5.242787e+09,0.000000,0.690000,0.000000
25%,101.000000,98111.000000,2.976195e+11,37.000000,2.197500,92.510000
50%,102.000000,98112.000000,5.091632e+11,50.000000,3.720000,190.400000
75%,103.000000,98113.000000,7.863253e+11,63.000000,9.747500,456.687500
max,103.000000,98113.000000,9.877695e+11,127.000000,24.510000,2254.920000


In [46]:
inventory[inventory['inventory_cost_value'] == 0].shape[0]

16

In [49]:
inventory[inventory['on_hand_qty'] == 0].shape[0]

16

In [50]:
inventory['store_name'].unique()

array(['Greenway - Northside', 'Greenway - Westlake',
       'Greenway - Riverside'], dtype=object)

In [51]:
inventory['department_name'].unique()

array(['Produce', 'Dairy', 'Bakery', 'Grocery', 'Wellness', 'Bulk',
       'Deli', 'Meat/Seafood'], dtype=object)

In [52]:
inventory['product_name'].unique()

array(['Rainshadow Produce Apples', 'Evergreen Farms Bananas',
       'Evergreen Farms Kale', 'Rainshadow Produce Spinach',
       'North Fork Organics Carrots', 'Rainshadow Produce Potatoes',
       'Rainshadow Produce Avocados', 'Rainshadow Produce Blueberries',
       'Rainshadow Produce Strawberries', 'Evergreen Farms Broccoli',
       'Skyline Dairy Milk 2%', 'Cascadia Creamery Whole Milk',
       'Cascadia Creamery Greek Yogurt',
       'Cascadia Creamery Butter Unsalted',
       'Cascadia Creamery Cheddar Cheese', 'EverFresh Cottage Cheese',
       'Cascadia Creamery Almond Milk', 'EverFresh Eggs',
       'Skyline Dairy Cream', 'EverFresh Sour Cream',
       'Blue Sky Bakery Sourdough Loaf', 'Stone Hearth Whole Wheat Bread',
       'Stone Hearth Croissant', 'Blue Sky Bakery Bagels',
       'Grain & Glow Blueberry Muffins', 'Sun Valley Rolled Oats',
       'Sun Valley Basmati Rice', 'Sun Valley Black Beans',
       'Sun Valley Spaghetti', 'Frontier Foods Olive Oil',
       'Sun V

## Inventory Data — Summary
* Number of Rows & Columns: 2160 samples, 13 features

* Key Columns Identified:
    * snapshot_date → Date of inventory record
    * store_id, store_name, store_city, store_state → Store details
    * product_upc, product_name, department_name → Product details
    * on_hand_qty → Quantity available
    * unit_cost → Cost per unit
    * inventory_cost_value → Total inventory value

* Missing Values: No missing values

  
* Duplicate Rows: No duplicate

  
* Observations:
    * Snapshot_date in other dtype.
    * Each record represents inventory of a product in a store on a specific date.
    * Store details repeat for multiple products.
    * Product UPC acts as a unique product identifier.
    * snapshot_date should be converted to datetime format later.
    * inventory_cost_value appears to be calculated as: on_hand_qty × unit_cost
    * Dataset is clean and well-structured for analysis.

### Product

In [53]:
product.head()

,product_upc,product_name,brand,department_name,category_name,size,unit,vendor_name,vendor_phone,regular_price,unit_cost,pack_size
0,433218196001,Rainshadow Produce Apples,Rainshadow Produce,Produce,Apples,1 lb,lb,Evergreen Organics,206-555-0110,3.11,2.31,1
1,386379402654,Evergreen Farms Bananas,Evergreen Farms,Produce,Bananas,1 lb,lb,Frontier Foods Wholesale,206-555-0118,5.33,3.68,1
2,161559407816,Evergreen Farms Kale,Evergreen Farms,Produce,Kale,3 lb,lb,Pacific Produce Co.,206-555-0111,1.76,1.10,1
3,931034131647,Rainshadow Produce Spinach,Rainshadow Produce,Produce,Spinach,3 lb,lb,Bulk Barn West,206-555-0115,1.15,0.84,6
4,534192832764,North Fork Organics Carrots,North Fork Organics,Produce,Carrots,2 lb,lb,Bulk Barn West,206-555-0115,4.52,3.19,6


In [54]:
product.shape

(60, 12)

In [55]:
product.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   product_upc      60 non-null     int64  
 1   product_name     60 non-null     object 
 2   brand            60 non-null     object 
 3   department_name  60 non-null     object 
 4   category_name    60 non-null     object 
 5   size             60 non-null     object 
 6   unit             60 non-null     object 
 7   vendor_name      60 non-null     object 
 8   vendor_phone     60 non-null     object 
 9   regular_price    60 non-null     float64
 10  unit_cost        60 non-null     float64
 11  pack_size        60 non-null     int64  
dtypes: float64(2), int64(2), object(8)
memory usage: 5.8+ KB


In [56]:
product.isnull().sum()

product_upc        0
product_name       0
brand              0
department_name    0
category_name      0
size               0
unit               0
vendor_name        0
vendor_phone       0
regular_price      0
unit_cost          0
pack_size          0
dtype: int64

In [57]:
product.duplicated().sum()

np.int64(0)

In [58]:
product.describe()

,product_upc,regular_price,unit_cost,pack_size
count,6.000000e+01,60.000000,60.0000,60.000000
mean,5.219156e+11,9.388333,6.3065,2.583333
std,2.911078e+11,8.281451,5.6051,2.010917
min,5.242787e+09,0.920000,0.6900,1.000000
25%,2.976195e+11,3.237500,2.1975,1.000000
50%,5.091632e+11,5.720000,3.7200,1.000000
75%,7.863253e+11,14.767500,9.7475,4.000000
max,9.877695e+11,35.340000,24.5100,6.000000


In [59]:
product['size'].unique()

array(['1 lb', '3 lb', '2 lb', 'Each', '12 ct', '1 gal', '1 qt', '16 oz',
       '8 oz', '32 oz', '1/2 gal', '24 oz', '6 ct', '48 oz', '12 oz',
       '120 ct', '60 ct', '90 ct', '0.5 lb', '1.5 lb'], dtype=object)

## Products Data- Summary
* Number of row & col: Sample 60, Features 12
  
* Key Column Identify:
    * product_upc -> unique no  
    * product_name, brand, department_name, category_name -> Product details
    * size -> Package Quantity
    * unit -> Measurement Type
    * vendor_name, vendor_phone -> Company detail
    * regular_price -> original price
    * unit_cost -> store purchase cost
    * pack_size -> How many items are inside one package
      
* Missing Values: 0

* Duplicated:- 0

* Observations:
    * Dataset contains 60 Productsample and 12 features
    * No duplicate records found
    * No Missing values
    * Avgerage 2 Pack_size
    * size column is messy

In [60]:
promotion.head()

,promo_id,product_upc,promo_type,discount_percent,start_date,end_date,product_name,department_name,brand
0,PR0011832,471349361832,PCT_OFF,15,2024-01-01,2024-01-07,Meadow Meats Salmon Fillet,Meat/Seafood,Meadow Meats
1,PR0019579,831727889579,LOYALTY_PCT,30,2024-01-01,2024-01-07,Sun Valley Tortillas,Grocery,Sun Valley
2,PR0013186,190229413186,LOYALTY_PCT,5,2024-01-01,2024-01-07,Greenway Kitchen Tomato Soup,Deli,Greenway Kitchen
3,PR0011904,452991241904,PCT_OFF,30,2024-01-01,2024-01-07,Sunrise Labs Magnesium,Wellness,Sunrise Labs
4,PR0015433,216073375433,BOGO50,50,2024-01-01,2024-01-07,Sun Valley Basmati Rice,Grocery,Sun Valley


In [61]:
promotion.shape

(530, 9)

In [62]:
promotion.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   promo_id          530 non-null    object
 1   product_upc       530 non-null    int64 
 2   promo_type        530 non-null    object
 3   discount_percent  530 non-null    int64 
 4   start_date        530 non-null    object
 5   end_date          530 non-null    object
 6   product_name      530 non-null    object
 7   department_name   530 non-null    object
 8   brand             530 non-null    object
dtypes: int64(2), object(7)
memory usage: 37.4+ KB


In [63]:
promotion.isnull().sum()

promo_id            0
product_upc         0
promo_type          0
discount_percent    0
start_date          0
end_date            0
product_name        0
department_name     0
brand               0
dtype: int64

In [64]:
promotion.duplicated().sum()

np.int64(0)

In [65]:
promotion.describe()

,product_upc,discount_percent
count,5.300000e+02,530.000000
mean,5.296574e+11,28.358491
std,2.984540e+11,17.162731
min,1.226917e+10,5.000000
25%,2.160734e+11,15.000000
50%,4.529912e+11,25.000000
75%,8.317279e+11,50.000000
max,9.513433e+11,50.000000


## Promotions Data — Summary
* Number of rows: 530 samples, 9 features
* Key Column Identify:
    * promo_id → Unique identifier for each promotion
    * product_upc → Unique product identifier linked to product table
    * promo_type → Type of promotion (e.g., PCT_OFF, BOGO50, LOYALTY_PCT)
    * discount_percent → Percentage discount applied in the promotion
    * start_date / end_date → Duration of the promotion period
    * product_name → Name of the product receiving the promotion
    * department_name → Product department category
    * brand → Brand associated with the product
    * Data Quality Checks

* Missing Values: 0
  
* Duplicate: 0

* Observations

    * Dataset contains 530 samples and 9 features.
    * No missing values detected across all columns.
    * No duplicate rows found in the dataset.
    * Promotions mainly range between 5% – 50% discount.
    * Date columns (start_date, end_date) are stored as object type, so they should be converted to datetime format for time-based analysis.
    * promo_type contains multiple promotion strategies (percentage discounts, loyalty discounts, buy-one-get-one offers).

### Sale

In [66]:
sale = pd.concat([sale_q1,sale_q2,sale_q3,sale_q4])

In [67]:
sale.head()

,receipt_id,line_number,sale_datetime,store_id,store_name,store_address,store_city,store_state,store_zip,cashier_name,tender_type,customer_segment,product_upc,product_name,brand,department_name,category_name,size,unit,vendor_name,vendor_phone,pack_size,regular_price,unit_cost,quantity,promo_type,promo_id,unit_price_effective,line_subtotal,tax_amount,weekend_flag
0,101-20240101-0001,1,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,566701065133,Cascadia Creamery Cheddar Cheese,Cascadia Creamery,Dairy,Cheddar Cheese,8 oz,oz,Ocean Crest Seafood,206-555-0116,1,5.34,3.24,1,NaN,NaN,5.34,5.34,0.0,False
1,101-20240101-0001,2,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,931034131647,Rainshadow Produce Spinach,Rainshadow Produce,Produce,Spinach,3 lb,lb,Bulk Barn West,206-555-0115,6,1.15,0.84,1,LOYALTY_PCT,PR0011647,1.09,1.09,0.0,False
2,101-20240101-0001,3,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,534192832764,North Fork Organics Carrots,North Fork Organics,Produce,Carrots,2 lb,lb,Bulk Barn West,206-555-0115,6,4.52,3.19,2,NaN,NaN,4.52,9.04,0.0,False
3,101-20240101-0001,4,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,455812236231,Sun Valley Chickpeas,Sun Valley,Grocery,Chickpeas,48 oz,oz,Blue Sky Bakery,206-555-0113,2,1.09,0.79,2,NaN,NaN,1.09,2.18,0.0,False
4,101-20240101-0002,1,01/01/2024 13:30,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Riley,Credit,Member,369957773872,Frontier Foods Peanut Butter,Frontier Foods,Grocery,Peanut Butter,48 oz,oz,Sunrise Wellness,206-555-0114,6,16.74,10.94,2,NaN,NaN,16.74,33.48,0.0,False


In [68]:
sale.info()

<class 'pandas.core.frame.DataFrame'>
Index: 53286 entries, 0 to 13164
Data columns (total 31 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   receipt_id            53286 non-null  object 
 1   line_number           53286 non-null  int64  
 2   sale_datetime         53286 non-null  object 
 3   store_id              53286 non-null  int64  
 4   store_name            53286 non-null  object 
 5   store_address         53286 non-null  object 
 6   store_city            53286 non-null  object 
 7   store_state           53286 non-null  object 
 8   store_zip             53286 non-null  int64  
 9   cashier_name          53286 non-null  object 
 10  tender_type           53286 non-null  object 
 11  customer_segment      53286 non-null  object 
 12  product_upc           53286 non-null  int64  
 13  product_name          53286 non-null  object 
 14  brand                 53286 non-null  object 
 15  department_name       53

In [69]:
sale.isnull().sum()

receipt_id                  0
line_number                 0
sale_datetime               0
store_id                    0
store_name                  0
store_address               0
store_city                  0
store_state                 0
store_zip                   0
cashier_name                0
tender_type                 0
customer_segment            0
product_upc                 0
product_name                0
brand                       0
department_name             0
category_name               0
size                        0
unit                        0
vendor_name                 0
vendor_phone                0
pack_size                   0
regular_price               0
unit_cost                   0
quantity                    0
promo_type              47991
promo_id                47991
unit_price_effective        0
line_subtotal               0
tax_amount                  0
weekend_flag                0
dtype: int64

In [70]:
sale.duplicated().sum()

np.int64(0)

In [71]:
sale.describe()

,line_number,store_id,store_zip,product_upc,pack_size,regular_price,unit_cost,quantity,unit_price_effective,line_subtotal,tax_amount
count,53286.000000,53286.000000,53286.000000,5.328600e+04,53286.000000,53286.000000,53286.000000,53286.000000,53286.000000,53286.000000,53286.000000
mean,2.989172,101.994614,98111.994614,5.274399e+11,2.482359,7.679584,5.185861,1.303851,7.546489,9.612246,0.129300
std,1.826760,0.818289,0.818289,2.942283e+11,1.950821,7.209329,4.911482,0.544938,7.109279,9.732685,0.583777
min,1.000000,101.000000,98111.000000,5.242787e+09,1.000000,0.920000,0.690000,1.000000,0.800000,0.800000,0.000000
25%,2.000000,101.000000,98111.000000,3.039117e+11,1.000000,2.680000,1.800000,1.000000,2.670000,3.110000,0.000000
50%,3.000000,102.000000,98112.000000,5.341928e+11,1.000000,4.520000,3.190000,1.000000,4.520000,5.500000,0.000000
75%,4.000000,103.000000,98113.000000,8.244935e+11,4.000000,12.340000,8.580000,2.000000,11.690000,13.250000,0.000000
max,13.000000,103.000000,98113.000000,9.877695e+11,6.000000,35.340000,24.510000,3.000000,35.340000,70.740000,3.570000


# STAGE 4 - Data Cleaning

In [72]:
# Inventory
inventory['snapshot_date'] = pd.to_datetime(inventory['snapshot_date'])

In [73]:
# Product
product['size'] = product['size'].replace('Each', '1 ct')

In [74]:
product[product['size'] == '1/2 gal']

,product_upc,product_name,brand,department_name,category_name,size,unit,vendor_name,vendor_phone,regular_price,unit_cost,pack_size
16,773602606474,Cascadia Creamery Almond Milk,Cascadia Creamery,Dairy,Almond Milk,1/2 gal,gal,Ocean Crest Seafood,206-555-0116,1.91,1.42,2


In [75]:
import re 

def extract_numeric(value):
    if isinstance(value, str):
        value = value.strip()
        
        # Case 1: Fraction (e.g., "1/2 gal")
        if re.search(r'\d+/\d+', value):
            frac = re.search(r'(\d+)/(\d+)', value)
            return float(frac.group(1)) / float(frac.group(2))
        
        # Case 2: Decimal or integer (e.g., "1.5 lb", "2 qt")
        num = re.search(r'\d+\.?\d*', value)
        if num:
            return float(num.group())
    
    return None

product['size'] = product['size'].apply(extract_numeric)

In [76]:
product.loc[16]

product_upc                         773602606474
product_name       Cascadia Creamery Almond Milk
brand                          Cascadia Creamery
department_name                            Dairy
category_name                        Almond Milk
size                                         0.5
unit                                         gal
vendor_name                  Ocean Crest Seafood
vendor_phone                        206-555-0116
regular_price                               1.91
unit_cost                                   1.42
pack_size                                      2
Name: 16, dtype: object

In [77]:
def categorize_unit(u):
    if u in ['lb', 'oz']:
        return 'Weight'
    elif u in ['gal', 'qt']:
        return 'Volume'
    elif u in ['ct']:
        return 'Count'
    else:
        return 'Other'

product['unit_category'] = product['unit'].apply(categorize_unit)

In [78]:
# promotion
promotion['start_date'] = pd.to_datetime(promotion['start_date'])

In [79]:
promotion['end_date'] = pd.to_datetime(promotion['end_date'])

In [80]:
promotion['duration'] = (promotion['end_date'] - promotion['start_date']).dt.days

In [81]:
# sale
sale['promo_type'].fillna('No Promo', inplace=True)
sale['promo_id'].fillna('No Promo', inplace=True)

C:\Users\HP\AppData\Local\Temp\ipykernel_2260\3557349377.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  sale['promo_type'].fillna('No Promo', inplace=True)
C:\Users\HP\AppData\Local\Temp\ipykernel_2260\3557349377.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example,

In [82]:
sale['size'] = sale['size'].replace('Each', '1 ct')

In [83]:
sale['size'] = sale['size'].apply(extract_numeric)

In [84]:
sale['unit_category'] = sale['unit'].apply(categorize_unit)

In [85]:
sale.head()

,receipt_id,line_number,sale_datetime,store_id,store_name,store_address,store_city,store_state,store_zip,cashier_name,tender_type,customer_segment,product_upc,product_name,brand,department_name,category_name,size,unit,vendor_name,vendor_phone,pack_size,regular_price,unit_cost,quantity,promo_type,promo_id,unit_price_effective,line_subtotal,tax_amount,weekend_flag,unit_category
0,101-20240101-0001,1,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,566701065133,Cascadia Creamery Cheddar Cheese,Cascadia Creamery,Dairy,Cheddar Cheese,8.0,oz,Ocean Crest Seafood,206-555-0116,1,5.34,3.24,1,No Promo,No Promo,5.34,5.34,0.0,False,Weight
1,101-20240101-0001,2,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,931034131647,Rainshadow Produce Spinach,Rainshadow Produce,Produce,Spinach,3.0,lb,Bulk Barn West,206-555-0115,6,1.15,0.84,1,LOYALTY_PCT,PR0011647,1.09,1.09,0.0,False,Weight
2,101-20240101-0001,3,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,534192832764,North Fork Organics Carrots,North Fork Organics,Produce,Carrots,2.0,lb,Bulk Barn West,206-555-0115,6,4.52,3.19,2,No Promo,No Promo,4.52,9.04,0.0,False,Weight
3,101-20240101-0001,4,01/01/2024 14:20,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Quinn,SNAP,Member,455812236231,Sun Valley Chickpeas,Sun Valley,Grocery,Chickpeas,48.0,oz,Blue Sky Bakery,206-555-0113,2,1.09,0.79,2,No Promo,No Promo,1.09,2.18,0.0,False,Weight
4,101-20240101-0002,1,01/01/2024 13:30,101,Greenway - Northside,1201 Maple Ave,Everton,WA,98111,Riley,Credit,Member,369957773872,Frontier Foods Peanut Butter,Frontier Foods,Grocery,Peanut Butter,48.0,oz,Sunrise Wellness,206-555-0114,6,16.74,10.94,2,No Promo,No Promo,16.74,33.48,0.0,False,Weight


In [86]:
sale['sale_datetime'] = pd.to_datetime(sale['sale_datetime'], format='mixed')

# STAGE 5 - Data Dictionary

The Data Dictionary documents the meaning, structure, and type of each column across the provided datasets. 
It helps ensure transparency and prepares the data for schema design and ETL implementation.

---

### 🏪 Inventory Dataset

| Column Name | Type | Description |
|--------------|------|-------------|
| snapshot_date | datetime | Date and time when the inventory snapshot was recorded |
| store_id | integer | Unique identifier for the store |
| store_name | text | Name of the store |
| store_address | text | Physical address of the store |
| store_city | text | City where the store is located |
| store_state | text | State where the store operates |
| store_zip | text | ZIP code of the store location |
| product_upc | integer | Unique product identifier |
| product_name | text | Name of the product |
| department_name | text | Department category of the product |
| on_hand_qty | integer | Quantity of product currently in stock |
| unit_cost | float | Cost of purchasing the product from the vendor |
| inventory_cost_value | float | Total inventory value (on_hand_qty × unit_cost) |

---

### 🧾 Sales Dataset

| Column Name | Type | Description |
|--------------|------|-------------|
| receipt_id | integer | Unique identifier for each sales receipt |
| line_number | integer | Line item number within the receipt |
| sale_datetime | datetime | Timestamp when the sale occurred |
| store_id | integer | Store where the transaction occurred |
| store_name | text | Name of the store |
| store_address | text | Address of the store |
| store_city | text | City where the store is located |
| store_state | text | State where the store operates |
| store_zip | text | ZIP code of the store location |
| cashier_name | text | Name of the cashier processing the transaction |
| tender_type | text | Payment method used (cash, card, etc.) |
| customer_segment | text | Customer classification segment |
| product_upc | integer | Unique identifier of the product sold |
| product_name | text | Name of the product |
| brand | text | Brand of the product |
| department_name | text | Department category |
| category_name | text | Product category |
| vendor_name | text | Supplier of the product |
| vendor_phone | text | Supplier contact number |
| pack_size | integer | Number of items in a product package |
| regular_price | float | Standard price of the product |
| unit_cost | float | Cost price for the store |
| quantity | integer | Number of units sold |
| promo_type | text | Type of promotion applied |
| promo_id | text | Identifier of the promotion campaign |
| unit_price_effective | float | Final price after discount |
| line_subtotal | float | Total price for the line item |
| tax_amount | float | Tax applied to the transaction |
| weekend_flag | boolean | Indicates whether sale occurred on weekend |
| numeric_size | float | Standardized product size |
| unit_type | text | Measurement unit |
| unit_category | text | Type of measurement (weight, volume, etc.) |

---

### 📦 Products Dataset

| Column Name | Type | Description |
|--------------|------|-------------|
| product_upc | integer | Unique identifier for the product |
| product_name | text | Name of the product |
| brand | text | Brand of the product |
| department_name | text | Department category |
| category_name | text | Product category |
| vendor_name | text | Supplier of the product |
| vendor_phone | text | Supplier contact information |
| regular_price | float | Standard selling price |
| unit_cost | float | Purchase cost for the store |
| pack_size | integer | Number of items in a product package |
| numeric_size | float | Standardized product size |
| unit_type | text | Measurement unit |
| unit_category | text | Type of measurement (weight, volume, etc.) |

---

### 🎯 Promotions Dataset

| Column Name | Type | Description |
|--------------|------|-------------|
| promo_id | text | Unique identifier for the promotion |
| product_upc | integer | Product associated with the promotion |
| promo_type | text | Type of promotion applied |
| discount_percent | float | Percentage discount applied |
| start_date | datetime | Promotion start date |
| end_date | datetime | Promotion end date |
| product_name | text | Name of the product |
| department_name | text | Department category |
| brand | text | Product brand |
| duration | integer | Number of days the promotion runs |

---



# STAGE 6 - ERD

![ERD](../Img/ERDiagram.png)

The raw datasets contain repeated information across multiple files such as store details, product details, vendor information, and transaction records. 
To remove redundancy and improve analytical usability, the data is organized into a conceptual relational model.

The following core entities were identified:

- Stores
- Products
- Vendors
- Transactions
- Line Items
- Promotions
- Inventory Snapshots

Each entity represents a key business component of the grocery retail system.

The ERD defines how these entities relate to one another and forms the foundation for the normalized relational database schema.

In [87]:
transactions = sale[[
    'receipt_id','store_id','sale_datetime',
    'cashier_name','tender_type','customer_segment'
]].drop_duplicates()

In [88]:
line_items = sale[[
    'receipt_id','product_upc', 'line_number',
    'quantity','unit_price_effective',
    'line_subtotal','tax_amount'
]]

In [99]:
stores = sale[[
    'store_id','store_name','store_address',
    'store_city','store_state','store_zip'
]].drop_duplicates()

In [90]:
products = sale[[
    'product_upc','product_name','brand',
    'department_name','category_name','size', 'unit',
    'pack_size','regular_price','unit_cost','vendor_name'
]].drop_duplicates()

In [91]:
vendors = sale[[
    'vendor_name','vendor_phone'
]].drop_duplicates()

In [92]:
inventory_df = inventory[[
    'snapshot_date','store_id','product_upc',
    'on_hand_qty','inventory_cost_value'
]]

In [106]:
inventory_df

,snapshot_date,store_id,product_upc,on_hand_qty,inventory_cost_value
0,2024-01-01,101,433218196001,35,80.85
1,2024-01-01,101,386379402654,43,158.24
2,2024-01-01,101,161559407816,42,46.20
3,2024-01-01,101,931034131647,20,16.80
4,2024-01-01,101,534192832764,55,175.45
...,...,...,...,...,...
2155,2024-12-01,103,471349361832,23,154.79
2156,2024-12-01,103,249947174648,46,782.46
2157,2024-12-01,103,906594013990,68,877.20
2158,2024-12-01,103,278742967175,47,621.81


In [93]:
promo = promotion[[
    'promo_id','product_upc','promo_type',
    'discount_percent','start_date','end_date','duration'
]]

In [94]:
vendors['vendor_id'] = range(1, len(vendors)+1)

In [95]:
products = products.merge(vendors, on='vendor_name', how='left')

In [96]:
products = products.drop(columns=['vendor_name','vendor_phone'])

In [ ]:
from sqlalchemy import create_engine

engine = create_engine(
    "postgresql://postgres:Guggli190502@localhost:5432/grocery_db"
)

: 

: 

: 

In [ ]:
inventory_df.to_sql('inventory_snapshots', engine, if_exists='append', index=False)

OperationalError: (psycopg2.OperationalError) connection to server at "localhost" (::1), port 5432 failed: FATAL:  database "grocery_db" does not exist

(Background on this error at: https://sqlalche.me/e/20/e3q8)

: 

: 

: 

In [ ]:
line_items.to_sql('line_items', engine, if_exists='append', index=False)

286

: 

: 

: 

: 

In [ ]:
promo.to_sql('promotions', engine, if_exists='append', index=False)

530

: 

: 

: 

: 

In [ ]:
products.to_sql('products', engine, if_exists='append', index=False)

60

: 

: 

: 

: 

In [ ]:
transactions.to_sql('transactions', engine, if_exists='append', index=False)
stores.to_sql('stores', engine, if_exists='append', index=False)
vendors.to_sql('vendors', engine, if_exists='append', index=False)

9

: 

: 

: 

: 

In [ ]:
import psycopg2

# Create connection
conn = psycopg2.connect(
    host="localhost",
    database="grocery_db",
    user="postgres",
    password="Guggli190502",
    port="5432"
)

print("Connected successfully!")

cur = conn.cursor()
cur.execute("select * from inventory_snapshots;")
rows = cur.fetchall()

for data in rows:
    print(data)

conn.close()


Connected successfully!
(datetime.datetime(2024, 1, 1, 0, 0), 101, 433218196001, 35, 80.85)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 386379402654, 43, 158.24)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 161559407816, 42, 46.2)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 931034131647, 20, 16.8)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 534192832764, 55, 175.45)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 30564139537, 55, 37.95)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 423884969653, 50, 188.0)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 12269166978, 45, 54.0)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 845146270482, 21, 38.22)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 932528809570, 41, 91.43)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 303911718227, 43, 146.2)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 963834657871, 33, 73.26)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 98393010310, 46, 146.74)
(datetime.datetime(2024, 1, 1, 0, 0), 101, 473829973763, 58, 109.62)
(datetime.datetime(2024, 1

: 

: 

: 

: 